internal/util/tempfiles.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+)
+
+// tempFiles is a slice of pointers to temporary files created throughout the course of the program. Users should call
+// RemoveTempFiles at the end of the program.
+var tempFiles []*os.File
+
+func CreateTempFile() (*os.File, error) {
```

Comment:
So this uses global state, and as such won't be threadsafe. 

A common pattern for this is to return a `closer` function that can cleanup after the resource is no longer needed. For example:
```go
tmp, cleanup, err := CreateTempFile()
if err != nil {

}
defer cleanup()
```
This is nice because you can defer it, making sure it gets cleaned up even if something goes wrong. It also helps scope the resource to exactly when it's needed in the function, instead of waiting for cleanup at the end. (and avoids the dangers of having a global state, like not being threadsafe etc). 

However, once you remove the `tempFiles` this is pretty much just the following:
```go
tempFile, err := ioutil.TempFile("", "example")
if err != nil {
	return tempFile, fmt.Errorf("[util.CreateTempFile] creating temp file: %w\n", err)
}
defer os.RemoveAll(tempFile)
```

So in that case a utility function may no longer be helpful.

Reply:
Resolved, thank you.